libc/src/link/linux/dl_iterate_phdr.cpp

```diff
@@ -0,0 +1,57 @@
+//===-- Implementation of dl_iterate_phdr ---------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/link/dl_iterate_phdr.h"
+#include "config/linux/app.h"
+#include "include/llvm-libc-macros/sys-auxv-macros.h"
+#include "src/__support/common.h"
+#include "src/__support/macros/config.h"
+
+[[gnu::weak,
+  gnu::visibility("hidden")]] extern const ElfW(Dyn) _DYNAMIC[]; // NOLINT
+
+#define AUX_CNT 38
+
+namespace LIBC_NAMESPACE_DECL {
+
+LLVM_LIBC_FUNCTION(int, dl_iterate_phdr,
```

Comment:
This looks a _lot_ like musl's `src/ldso/dl_iterate_phdr.c`. If that's the case, please provide some attribution, and also mention it in the commit message.

Reply:
Will do

Reply:
Simply mentioning it comes from musl isn't sufficient -- musl's COPYRIGHT file imposes very specific requirements, as does LLVM's own license file:

https://github.com/llvm/llvm-project/blob/814902a03a2bb4114fd61c05e0e599fc98d61dbb/LICENSE.TXT#L224-L234

MIT has very few requirements but what it does require has not been satisfied by the most recent push.

Reply:
*Fun*, I've not really had to deal with licensing things like this before. I'm not sure what exactly to do.

Reply:
Please research the topic before making further PRs to (any) projects -- it's important that you understand licencing and using code derived from another project without honouring their licence. I don't think I can explain the whole topic in a succinct comment here but there's a lot of resources online about it.

Reply:
This is the only time I've copied other implementations like this. I tried figuring out what `dl_iterate_phdrs` but I didn't really understand it from the man page. Is there a way that this implementation could be changed enough that it then doesn't have a licensing problem?

Reply:
The legal threshold for changing something enough that it then doesn't have a licensing problem is:

> don't base the code on something with a licensing problem, understand the problem you are trying to solve well enough to write a clean-room implementation by hand, yourself, from scratch, without the aid of another implementation

But like I said, llvm's own LICENSE.TXT and musl's COPYRIGHT provide guidance on how to respect the license for a copy-pasted work (as long as the the terms of each license don't pose an onerous burden that the other license cannot abide by).

Reply:
I'm not sure how that works. Musl is MIT, LLVM is Apache. I've tried reading the licenses but I don't understand them.